v4.1.0

airon-applyinnovations opened the pull request:
# Created hooks that uses apollo to connect with appsync.
- supports query
- supports mutation
- supports subscriptions
- supports refetch on mutation events
- mostly typed. e.g. by passing query/hook/subscription names you will be guided on what the needed fields you need to pass and what the results will be
- script that generates additional type base on `api.ts` file. `yarn typegen ./path/to/api.ts`. `typegen` will be added to `node_modules/.bin` folder on install
- one setup function to set types, queries, mutations, subscriptions. this setup function will create the hooks and providers. the created hooks and providers will be aware of the types. when using them you won't need to pass the types anymore. 

## Acceptance Criteria

 - [x] Supports querying all queries and mutations hosted in the GraphQL schema
- [x] All inputs and responses are strongly typed using 100% codegen  (WILL NEED TO RUN `yarn typegen common/api.ts` after codegen for additional type to be generated)
- [x] This implementation should prioritize developer experience (after configuring the needed types, queries, mutations, subscriptions the hooks don't need additional typing)


## Screenshots
- Sample use of setup function
![image](https://user-images.githubusercontent.com/111259764/192261010-03ef174a-927d-4c1a-8d8c-c42464e07376.png)


- Configurable to refetch queries on mutation events. typed, autocomplete will work while doing the configuration
![Screen Shot 2022-09-26 at 6 49 04 PM](https://user-images.githubusercontent.com/111259764/192258697-6840a325-0a3f-46c2-893a-cddde0a81fc3.jpg)

- Inputs are typed as well as file fields
![Screen Shot 2022-09-26 at 5 14 59 PM](https://user-images.githubusercontent.com/111259764/192258705-b7320b7b-7c8c-49b6-993e-05facc4a3adc.jpg)
![Screen Shot 2022-09-26 at 6 22 54 PM](https://user-images.githubusercontent.com/111259764/192258702-a1130d4f-dacd-4fd5-95ea-d15e123e5b6f.jpg)
![Screen Shot 2022-09-26 at 5 13 21 PM](https://user-images.githubusercontent.com/111259764/192258710-940cf452-b97e-47c3-869b-1dc636d81fb2.jpg)

- autocomplete mutation/query/subscription names
![Screen Shot 2022-09-26 at 5 14 02 PM](https://user-images.githubusercontent.com/111259764/192258707-cc4bba70-5522-407c-984c-62be065224c0.jpg)
![Screen Shot 2022-09-26 at 5 12 47 PM](https://user-images.githubusercontent.com/111259764/192258713-e68e75ae-e168-46e0-ac3c-5860d0803f6b.jpg)

- typed fields of result data
![image](https://user-images.githubusercontent.com/111259764/192263377-5de8a498-70f8-4adf-97b0-e93fbf6ccfc9.png)

Commit:
added appsync hooks

Commit:
added appsync hooks

Commit:
update types

Commit:
types

Commit:
types

Commit:
types

Commit:
typegen

Commit:
brought back subscription refetch

actions-user committed:
[**/dist/**] generated build files; [package.json] updated to ;

airon-applyinnovations commented on packages/appsync/src/createAppSyncHooks.tsx:
```diff
@@ -0,0 +1,198 @@
+  mutations: Record<keyof MT, string>;
+  subscriptions: Record<keyof ST, string>;
+  url: string;
+  type: "AMAZON_COGNITO_USER_POOLS";
```
did not bother for other auth provider for now but can be extended in the future if we need other providers.

airon-applyinnovations commented on packages/appsync/src/createAppSyncHooks.tsx:
```diff
@@ -0,0 +1,198 @@
+    return null;
+  };
+
+  const RefetchSubscriptions: React.FC = () => {
```
because hooks can't be inside a loop I had to create components that calls useSubscription but returns null instead. this is for the refetch functionality that listens to subscription events.

airon-applyinnovations commented on packages/appsync/src/storageUtils.ts:
```diff
@@ -0,0 +1,110 @@
+import { Auth } from "@aws-amplify/auth";
```
code from datastore hook

airon-applyinnovations commented on packages/appsync/src/types.ts:
```diff
@@ -0,0 +1,40 @@
+import { StorageAccessLevel } from "@aws-amplify/storage";
```
code based from datastore hook with some modification

airon-applyinnovations commented on packages/appsync/src/useStorageObject.tsx:
```diff
@@ -0,0 +1,71 @@
+import { useEffect, useRef, useState } from "react";
```
code from datastore hook with some modification

airon-applyinnovations commented on packages/auth/src/AuthProvider.tsx:
```diff
@@ -195,6 +195,17 @@ export const AuthProvider: React.FC<AuthProviderProps> = ({
 
   const getPassword = (phoneNumber: string) => MD5(`${phoneNumber}`).toString();
 
+  const signInUser = useCallback(
```
this was remove on previous update of this repo but I realized we don't want adminweb to use signInOrCreate.

Commit:
reverted change to datastore

Commit:
reverted change to datastore

Commit:
reverted change to datastore

Commit:
reverted change to datastore

Commit:
react 17; forcing new commit

alexander-applyinnovations commented:
In relation to the auth provider situation, would it just be simpler to pass a token directly to the provider component instead of coupling the auth logic into this package?

peter-applyinnovations commented on packages/appsync/src/createAppSyncHooks.tsx:
```diff
@@ -0,0 +1,198 @@
+        ...opts,
+        variables: {
+          input: {
+            ...joinedInput,
```
you can just directly pass this

input: joinedInput

airon-applyinnovations replied:
updated

peter-applyinnovations reviewed (changes requested):
Looks ok to me, just left a minor comment

Commit:
extract auth from appsync

Commit:
remove unecessary destructure

actions-user committed:
[**/dist/**] generated build files; [package.json] updated to v4.1.0;

airon-applyinnovations commented:
> In relation to the auth provider situation, would it just be simpler to pass a token directly to the provider component instead of coupling the auth logic into this package?

Updated
- removed provider that handles authentication
- updated GraphqlProvider to accept apollo client. accepting token may not be enough because url, region and auth option (IAM,oAuth,APIkey,None,Lambda) needs to be pass to client. 
- made buildClient exportable so package user may use it to easily build the client with custom auth options they want.

Commit:
force build

alexander-applyinnovations commented on packages/appsync/bin/typegen:
```diff
@@ -0,0 +1,141 @@
+
+const FILE_PATH = path.join(process.cwd(), apiPath);
+
+const QUERY_REGEX = /export type (((List|Get|Sync).*)Query) =/;
```
This would not work for custom queries, mutations and subscriptions. Perhaps we should extract the names from the graphql schema and then extract them accordingly from the codegenerated types files?

alexander-applyinnovations replied:
maybe we should do this as a separate task tho

airon-applyinnovations replied:
yes this won't work if the custom queries don't follow the same naming format. we can do improvements on it on a separate task.

Commit:
fix conflict

airon-applyinnovations commented:
closing inactive PR
